### PR TITLE
Data cleanup (towards source_factors PR)

### PR DESCRIPTION
**Pull request by fhieber**
_This is a PR against the source factors branch_ to include another backwards-incompatible change:
Separates DataConfig into DataConfig and DataInfo.
DataInfo contains only stuff that is relevant for training (and prepared data folders). It is saved to a file called `data.info`.
DataConfig is part of ModelConfig and contains values relevant for inference.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./pre-commit.sh` or manual run of pylint & mypy)
- [x] You have considered writing a test
- [ ] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [ ] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/data_io.py [hunk 1]
```diff
@@ -556,13 +556,17 @@ def prepare_data(source_fnames: List[str],
...
-                             num_shards=num_shards,
-                             data_statistics=data_statistics,
+    data_info = DataInfo(sources=[os.path.abspath(fname) for fname in source_fnames],
+                         target=os.path.abspath(target_fname),
+                         source_vocabs=source_vocab_paths,
+                         target_vocab=target_vocab_path,
+                         shared_vocab=shared_vocab,
+                         num_shards=num_shards)
+    data_info_fname = os.path.join(output_prefix, C.DATA_INFO)
+    logger.info("Writing data config to '%s'", data_info_fname)
```
Comment (tdomhan) [inline sockeye/data_io.py]: technically "writing data info..." ;)

File: sockeye/data_io.py [hunk 2]
```diff
@@ -909,21 +919,30 @@ def __init__(self,
...
+
+    def __init__(self,
+                 data_statistics: DataStatistics,
+                 max_seq_len_source: int,
+                 max_seq_len_target: int) -> None:
+        super().__init__()
         self.data_statistics = data_statistics
         self.max_seq_len_source = max_seq_len_source
         self.max_seq_len_target = max_seq_len_target
-        self.num_source_factors = len(sources)
```
Comment (tdomhan) [inline sockeye/data_io.py]: wouldn't it make sense to keep the num_source_factors here, as it's something that is necessary at inference time?

File: sockeye/data_io.py (hunk 1 above)
Comment (fhieber) [reply sockeye/data_io.py]: fixed

File: sockeye/data_io.py (hunk 2 above)
Comment (fhieber) [reply sockeye/data_io.py]: At inference time we currently get them from the embedding config:
```
@property
    def num_source_factors(self) -> int:
        """
        Returns the number of source factors of this InferenceModel (at least 1).
        """
        return self.config.config_embed_source.num_factors
```
Which one do you prefer?

File: sockeye/data_io.py [hunk 3]
```diff
@@ -898,9 +908,9 @@ def describe_data_and_buckets(data_statistics: DataStatistics, bucket_batch_size
                         bucket_batch_size.average_words_per_batch)
 
 
-class DataConfig(config.Config):
+class DataInfo(config.Config):
     """
-    Stores data paths from training.
+    Stores training data information that is not relevant for inference
```
Comment (davvil) [inline sockeye/data_io.py]: Missing period at the end.

File: sockeye/data_io.py (hunk 2 above)
Comment (davvil) [reply sockeye/data_io.py]: I think it is cleaner in DataConfig, but it doesn't really make a difference...

Comment (fhieber) [reply sockeye/data_io.py]: agreed, I changed it

File: sockeye/train.py [hunk 4]
```diff
@@ -604,7 +607,7 @@ def create_model_config(args: argparse.Namespace,
     config_decoder = create_decoder_config(args, encoder_num_hidden)
 
     source_factor_configs = None
-    if config_data.num_source_factors > 1:
+    if len(source_vocab_sizes) > 1:
```
Comment (tdomhan) [inline sockeye/train.py]: it probably makes sense to add a check for the args.source_factors_num_embed being of equal size to source_factor_vocab_sizes.

**Pull request merged**